File: snippets/cpp/09_dynamic_programming/longest_common_subsequence.cpp
#include <algorithm>
#include <string>
#include <vector>

int lcs_length(const std::string& a, const std::string& b) {
    std::vector<std::vector<int>> dp(
        a.size() + 1,
        std::vector<int>(b.size() + 1, 0)
    );

    for (size_t i = 1; i <= a.size(); ++i) {
        for (size_t j = 1; j <= b.size(); ++j) {
            if (a[i - 1] == b[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = std::max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    return dp[a.size()][b.size()];
}

File: snippets/rust/09_dynamic_programming/longest_common_subsequence.rs
pub fn lcs_length(a: &str, b: &str) -> usize {
    let a_chars: Vec<char> = a.chars().collect();
    let b_chars: Vec<char> = b.chars().collect();
    let mut dp = vec![vec![0usize; b_chars.len() + 1]; a_chars.len() + 1];

    for i in 1..=a_chars.len() {
        for j in 1..=b_chars.len() {
            if a_chars[i - 1] == b_chars[j - 1] {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = dp[i - 1][j].max(dp[i][j - 1]);
            }
        }
    }

    dp[a_chars.len()][b_chars.len()]
}
